chain/filecoin/account.go

```diff
@@ -0,0 +1,294 @@
+	}
+	if len(tx.msg.Params) == 0 {
+		return contract.CallData([]byte{byte(tx.msg.Method)})
+	}
+	return contract.CallData(append([]byte{byte(tx.msg.Method)}, tx.msg.Params...))
+}
+
+// Sighashes returns the digests that must be signed before the transaction
+// can be submitted by the client.
+func (tx Tx) Sighashes() ([]pack.Bytes, error) {
+	sighash32 := blake2b.Sum256(tx.Hash())
+	sighash := sighash32[:]
+	return []pack.Bytes{pack.NewBytes(sighash)}, nil
+}
+
+// Sign the transaction by injecting signatures for the required sighashes.
+// The serialized public key used to sign the sighashes must also be
+// specified.
+func (tx *Tx) Sign(signatures []pack.Bytes, pubkey pack.Bytes) error {
+	if len(signatures) != 1 {
+		return fmt.Errorf("expected 1 signature, got %v signatures", len(signatures))
+	}
+
+	if len(signatures[0]) != 65 {
+		return fmt.Errorf("expected signature to be 65 bytes, got %v bytes", len(signatures[0]))
+	}
+
+	// we are sure there are exactly 65 bytes in the signature, so this is safe
+	sig := [65]byte{}
+	for i := range sig {
+		sig[i] = signatures[0][i]
+	}
+
+	tx.signature = pack.NewBytes65(sig)
+	return nil
+}
+
+// Serialize the transaction into bytes. Generally, this is the format in
+// which the transaction will be submitted by the client.
+func (tx Tx) Serialize() (pack.Bytes, error) {
+	buf := new(bytes.Buffer)
+	if err := tx.msg.MarshalCBOR(buf); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// TxBuilder represents a transaction builder that builds transactions to be
+// broadcasted to the filecoin network. The TxBuilder is configured using a
+// gas price and gas limit.
+type TxBuilder struct {
+	gasFeeCap  pack.U256
+	gasLimit   pack.U256
+	gasPremium pack.U256
+}
+
+// NewTxBuilder creates a new transaction builder.
+func NewTxBuilder(gasFeeCap, gasLimit, gasPremium pack.U256) TxBuilder {
+	return TxBuilder{gasFeeCap: gasFeeCap, gasLimit: gasLimit, gasPremium: gasPremium}
+}
+
+// BuildTx receives transaction fields and constructs a new transaction.
+func (txBuilder TxBuilder) BuildTx(from, to address.Address, value, nonce, _, _ pack.U256, payload pack.Bytes) (account.Tx, error) {
+	filfrom, err := filaddress.NewFromString(string(from))
+	if err != nil {
+		return nil, fmt.Errorf("bad from address '%v': %v", from, err)
+	}
+	filto, err := filaddress.NewFromString(string(to))
+	if err != nil {
+		return nil, fmt.Errorf("bad to address '%v': %v", to, err)
+	}
+	methodNum := abi.MethodNum(0)
+	if len(payload) > 0 {
+		methodNum = abi.MethodNum(payload[0])
+		payload = payload[1:]
+	}
+	return &Tx{
+		msg: types.Message{
+			Version:    types.MessageVersion,
+			From:       filfrom,
+			To:         filto,
+			Value:      big.Int{Int: value.Int()},
+			Nonce:      nonce.Int().Uint64(),
+			GasFeeCap:  big.Int{Int: txBuilder.gasFeeCap.Int()},
+			GasLimit:   txBuilder.gasLimit.Int().Int64(),
+			GasPremium: big.Int{Int: txBuilder.gasPremium.Int()},
```

Comment:
It makes sense for the `gasPremium` to be part of the `TxBuilder` constructor, but the `gasFeeCap` and `gasFeeLimit` should come from the inputs to the `BuildTx` function. Right now, those inputs are simply being ignored, but we could just use them instead. In this case, the Multichain can use the limit/price to infer the limit/cap (and then the premium can come from the constructor, as it does right now).